
# About the data:

## The data is about the waiting time in a hospital.
## The problem statement is: the clinic has gotten several complaints regarding wait times.
 ____________
## -- The questions from the management about the data are:
### Dose the patient type affect the waiting time?
### Is there a specific type of patient waiting a long time?
### Are we too busy?
### Do we have staffing issues?
### How muche the patients wait before the doctor can see them?
### What type of staff do we need or where do we need them?
### What days of the week are affected?
### How can we fix it?

In [1]:
import pandas as pp
import numpy as nn
import plotly.express as px
import plotly.offline as py
template_style = "plotly_dark"
%matplotlib inline

In [2]:
df = pp.read_excel(r'C:\" File path "\hospital_data_sampleee.xlsx')      #read the file
df.head()

Date  Medication Revenue    Lab Cost   Consultation Revenue   \
0 2019-11-04              1183.22          10                  20.17   
1 2019-11-06               738.48       $-                        15   
2 2019-11-02                  660       $-                     21.17   
3 2019-11-06                  600       $-                     $-      
4 2019-11-01                591.6       $-                        12   

  Doctor Type Financial Class Patient Type Entry Time Post-Consultation Time  \
0      ANCHOR             HMO   OUTPATIENT   08:35:45               09:17:54   
1      ANCHOR       INSURANCE   OUTPATIENT   19:19:16               21:02:36   
2      ANCHOR             HMO   OUTPATIENT   10:46:52               11:56:25   
3      ANCHOR        MEDICARE   OUTPATIENT   09:38:34               10:55:50   
4      ANCHOR       INSURANCE   OUTPATIENT   11:16:21               12:06:49   

  Completion Time Patient ID  
0        09:29:46     C10001  
1        21:24:07     C10002  
2        12:06:28     C10003  
3        10:58:02     C10004  
4        12:06:54     C10005

## EDA/CLEANING

In [3]:
df.info()       #the information of the data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29998 entries, 0 to 29997
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    29998 non-null  datetime64[ns]
 1    Medication Revenue     29998 non-null  object        
 2     Lab Cost              29998 non-null  object        
 3    Consultation Revenue   29998 non-null  object        
 4   Doctor Type             29998 non-null  object        
 5   Financial Class         29998 non-null  object        
 6   Patient Type            29998 non-null  object        
 7   Entry Time              29998 non-null  object        
 8   Post-Consultation Time  29998 non-null  object        
 9   Completion Time         29998 non-null  object        
 10  Patient ID              29998 non-null  object        
dtypes: datetime64[ns](1), object(10)
memory usage: 2.5+ MB


In [4]:
df.shape

(29998, 11)

In [5]:
df.duplicated().sum()           #Is there any duplicated values?

0

In [6]:
df.isnull().sum().sort_values(ascending=False)          #is there any NULL values?

Date                      0
 Medication Revenue       0
  Lab Cost                0
 Consultation Revenue     0
Doctor Type               0
Financial Class           0
Patient Type              0
Entry Time                0
Post-Consultation Time    0
Completion Time           0
Patient ID                0
dtype: int64

In [7]:
df.nunique()

Date                         13
 Medication Revenue        4280
  Lab Cost                  200
 Consultation Revenue       241
Doctor Type                   3
Financial Class               5
Patient Type                  1
Entry Time                20628
Post-Consultation Time    21356
Completion Time           21384
Patient ID                29998
dtype: int64

In [8]:
df_copy = df.copy()         # always make sure u keep a copy

In [9]:
df.columns = df.columns.str.strip().str.lower().str.replace(" ", "_").str.replace("-","_" )       ## change the names forms

In [10]:
df.head()

date medication_revenue lab_cost consultation_revenue doctor_type  \
0 2019-11-04            1183.22       10                20.17      ANCHOR   
1 2019-11-06             738.48    $-                      15      ANCHOR   
2 2019-11-02                660    $-                   21.17      ANCHOR   
3 2019-11-06                600    $-                   $-         ANCHOR   
4 2019-11-01              591.6    $-                      12      ANCHOR   

  financial_class patient_type entry_time post_consultation_time  \
0             HMO   OUTPATIENT   08:35:45               09:17:54   
1       INSURANCE   OUTPATIENT   19:19:16               21:02:36   
2             HMO   OUTPATIENT   10:46:52               11:56:25   
3        MEDICARE   OUTPATIENT   09:38:34               10:55:50   
4       INSURANCE   OUTPATIENT   11:16:21               12:06:49   

  completion_time patient_id  
0        09:29:46     C10001  
1        21:24:07     C10002  
2        12:06:28     C10003  
3        10:58:02     C10004  
4        12:06:54     C10005

### let's look at every single columns

In [11]:
df.date.value_counts().sort_index()

2019-11-01    2518
2019-11-02    1471
2019-11-03    1301
2019-11-04    3365
2019-11-05    2798
2019-11-06    2813
2019-11-07    2673
2019-11-08    2405
2019-11-09    1539
2019-11-10    1248
2019-11-11    3617
2019-11-12    2892
2019-11-13    1358
Name: date, dtype: int64

In [12]:
df.medication_revenue.value_counts()

 $-       11936
7.48        149
5           129
17          126
8.5         118
          ...  
56.62         1
56.6          1
56.52         1
56.5          1
41.34         1
Name: medication_revenue, Length: 4280, dtype: int64

In [13]:
df.lab_cost.value_counts()

 $-       28565
10          191
8           138
28.04        96
15           90
          ...  
37            1
48.15         1
107           1
101           1
9.5           1
Name: lab_cost, Length: 200, dtype: int64

In [14]:
df.consultation_revenue.value_counts()

 $-       5576
13        4762
12        3158
15        2509
24.3      1033
          ... 
23.36        1
52           1
35.13        1
17.76        1
14.5         1
Name: consultation_revenue, Length: 241, dtype: int64

In [15]:
df.doctor_type.value_counts()

ANCHOR      21913
LOCUM        6789
FLOATING     1296
Name: doctor_type, dtype: int64

In [16]:
df.financial_class.value_counts()

INSURANCE    9931
PRIVATE      9121
CORPORATE    6915
HMO          3738
MEDICARE      293
Name: financial_class, dtype: int64

In [17]:
df.patient_type.value_counts()

OUTPATIENT    29998
Name: patient_type, dtype: int64

In [18]:
df.entry_time.value_counts()

18:07:47    9
18:02:40    7
10:29:55    7
08:39:21    7
10:59:54    7
           ..
18:53:26    1
14:23:33    1
16:08:18    1
16:29:46    1
13:58:57    1
Name: entry_time, Length: 20628, dtype: int64

In [19]:
df.post_consultation_time.value_counts()

11:44:23    6
10:32:26    6
11:04:16    6
11:45:35    6
11:11:20    6
           ..
21:05:23    1
13:09:24    1
14:19:54    1
18:40:57    1
14:13:21    1
Name: post_consultation_time, Length: 21356, dtype: int64

In [20]:
df.completion_time.value_counts()

10:04:33    7
10:12:54    7
10:20:45    6
10:04:23    6
09:51:23    6
           ..
18:38:03    1
11:31:34    1
12:55:16    1
09:02:37    1
14:19:02    1
Name: completion_time, Length: 21384, dtype: int64

In [21]:
df.patient_id.value_counts()

C10001    1
C29995    1
C30007    1
C30006    1
C30005    1
         ..
C19996    1
C19995    1
C19994    1
C19993    1
C39999    1
Name: patient_id, Length: 29998, dtype: int64

### (EDA)

In [22]:
df.dtypes

date                      datetime64[ns]
medication_revenue                object
lab_cost                          object
consultation_revenue              object
doctor_type                       object
financial_class                   object
patient_type                      object
entry_time                        object
post_consultation_time            object
completion_time                   object
patient_id                        object
dtype: object

### We don't have the data for patient time waiting, so let's add the timing by sub 'entry time' - 'completion time'

### We cant sub two dates.time so we need to change 'entry_time' , 'post_consultation_time' , 'completion_time' to_timedelta

In [23]:
df.completion_time = pp.to_timedelta(df.completion_time.astype(str))
df.entry_time = pp.to_timedelta(df.entry_time.astype(str))
df.post_consultation_time = pp.to_timedelta(df.post_consultation_time.astype(str))

In [24]:
df.dtypes

date                       datetime64[ns]
medication_revenue                 object
lab_cost                           object
consultation_revenue               object
doctor_type                        object
financial_class                    object
patient_type                       object
entry_time                timedelta64[ns]
post_consultation_time    timedelta64[ns]
completion_time           timedelta64[ns]
patient_id                         object
dtype: object

In [25]:
df

date medication_revenue lab_cost consultation_revenue doctor_type  \
0     2019-11-04            1183.22       10                20.17      ANCHOR   
1     2019-11-06             738.48    $-                      15      ANCHOR   
2     2019-11-02                660    $-                   21.17      ANCHOR   
3     2019-11-06                600    $-                   $-         ANCHOR   
4     2019-11-01              591.6    $-                      12      ANCHOR   
...          ...                ...      ...                  ...         ...   
29993 2019-11-13              $-       $-                   $-         ANCHOR   
29994 2019-11-13              $-       $-                   $-         ANCHOR   
29995 2019-11-13              $-       $-                   $-         ANCHOR   
29996 2019-11-13              $-       $-                   $-          LOCUM   
29997 2019-11-13              $-       $-                   $-         ANCHOR   

      financial_class patient_type      entry_time post_consultation_time  \
0                 HMO   OUTPATIENT 0 days 08:35:45        0 days 09:17:54   
1           INSURANCE   OUTPATIENT 0 days 19:19:16        0 days 21:02:36   
2                 HMO   OUTPATIENT 0 days 10:46:52        0 days 11:56:25   
3            MEDICARE   OUTPATIENT 0 days 09:38:34        0 days 10:55:50   
4           INSURANCE   OUTPATIENT 0 days 11:16:21        0 days 12:06:49   
...               ...          ...             ...                    ...   
29993         PRIVATE   OUTPATIENT 0 days 14:01:24        0 days 14:08:56   
29994         PRIVATE   OUTPATIENT 0 days 14:01:33        0 days 14:21:17   
29995         PRIVATE   OUTPATIENT 0 days 14:01:41        0 days 14:13:57   
29996       CORPORATE   OUTPATIENT 0 days 14:01:42        0 days 14:13:21   
29997         PRIVATE   OUTPATIENT 0 days 14:01:55        0 days 14:18:40   

      completion_time patient_id  
0     0 days 09:29:46     C10001  
1     0 days 21:24:07     C10002  
2     0 days 12:06:28     C10003  
3     0 days 10:58:02     C10004  
4     0 days 12:06:54     C10005  
...               ...        ...  
29993 0 days 14:15:42     C39995  
29994 0 days 14:21:53     C39996  
29995 0 days 14:16:29     C39997  
29996 0 days 14:13:22     C39998  
29997 0 days 14:19:02     C39999  

[29998 rows x 11 columns]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29998 entries, 0 to 29997
Data columns (total 11 columns):
 #   Column                  Non-Null Count  Dtype          
---  ------                  --------------  -----          
 0   date                    29998 non-null  datetime64[ns] 
 1   medication_revenue      29998 non-null  object         
 2   lab_cost                29998 non-null  object         
 3   consultation_revenue    29998 non-null  object         
 4   doctor_type             29998 non-null  object         
 5   financial_class         29998 non-null  object         
 6   patient_type            29998 non-null  object         
 7   entry_time              29998 non-null  timedelta64[ns]
 8   post_consultation_time  29998 non-null  timedelta64[ns]
 9   completion_time         29998 non-null  timedelta64[ns]
 10  patient_id              29998 non-null  object         
dtypes: datetime64[ns](1), object(7), timedelta64[ns](3)
memory usage: 2.5+ MB


In [27]:
df.head(2)

date medication_revenue lab_cost consultation_revenue doctor_type  \
0 2019-11-04            1183.22       10                20.17      ANCHOR   
1 2019-11-06             738.48    $-                      15      ANCHOR   

  financial_class patient_type      entry_time post_consultation_time  \
0             HMO   OUTPATIENT 0 days 08:35:45        0 days 09:17:54   
1       INSURANCE   OUTPATIENT 0 days 19:19:16        0 days 21:02:36   

  completion_time patient_id  
0 0 days 09:29:46     C10001  
1 0 days 21:24:07     C10002

#### just ignore the 0 days

### We just need the 'waiting_time', but ill add some more to make it easy for me

In [28]:
df['waiting_time'] = df['completion_time'] - df['entry_time']  # new column for Calculator 'waiting_time'
###
df['waiting_ber_munets'] = df.waiting_time.dt.seconds / 60     # new column for Calculator 'waiting_ber_munets'
df['waiting_ber_munets'] = df['waiting_ber_munets'].round(0)
####
df['weekday'] = df.date.dt.strftime('%A')                      # new column for extract 'weekday' from 'date'
df['hours'] = df.entry_time.dt.components.hours                # new column for extract the  'hours' from 'entry_time'
df.head()

date medication_revenue lab_cost consultation_revenue doctor_type  \
0 2019-11-04            1183.22       10                20.17      ANCHOR   
1 2019-11-06             738.48    $-                      15      ANCHOR   
2 2019-11-02                660    $-                   21.17      ANCHOR   
3 2019-11-06                600    $-                   $-         ANCHOR   
4 2019-11-01              591.6    $-                      12      ANCHOR   

  financial_class patient_type      entry_time post_consultation_time  \
0             HMO   OUTPATIENT 0 days 08:35:45        0 days 09:17:54   
1       INSURANCE   OUTPATIENT 0 days 19:19:16        0 days 21:02:36   
2             HMO   OUTPATIENT 0 days 10:46:52        0 days 11:56:25   
3        MEDICARE   OUTPATIENT 0 days 09:38:34        0 days 10:55:50   
4       INSURANCE   OUTPATIENT 0 days 11:16:21        0 days 12:06:49   

  completion_time patient_id    waiting_time  waiting_ber_munets    weekday  \
0 0 days 09:29:46     C10001 0 days 00:54:01                54.0     Monday   
1 0 days 21:24:07     C10002 0 days 02:04:51               125.0  Wednesday   
2 0 days 12:06:28     C10003 0 days 01:19:36                80.0   Saturday   
3 0 days 10:58:02     C10004 0 days 01:19:28                79.0  Wednesday   
4 0 days 12:06:54     C10005 0 days 00:50:33                51.0     Friday   

   hours  
0      8  
1     19  
2     10  
3      9  
4     11

### The first thing they want to know is: Dose the patient type affect the waiting time? , And is there a specific type of patients waiting a long time?

In [29]:
df.patient_type.value_counts()          #checking the values of 'patient_type'

OUTPATIENT    29998
Name: patient_type, dtype: int64

In [30]:
df.financial_class.value_counts()

INSURANCE    9931
PRIVATE      9121
CORPORATE    6915
HMO          3738
MEDICARE      293
Name: financial_class, dtype: int64

### We just have one type of patient in the 'patient_type', but if we can use 'financial_class' I think we can answer the Q

#### And we will do the same for 'Weekday' later (the second Q) , so let's build a function for Groupby

In [31]:

# Groupby as a function
def grouped_data(column_name):
    '''
    Groupby column and return DataFrame
    Input: Column Name
    '''
    df_tmp = df.groupby(column_name)[['waiting_ber_munets']].mean().round(0)
    df_tmp2 = df[column_name].value_counts()      #this for concat
    
    pivot_F = pp.concat([df_tmp, df_tmp2.rename('number_of_patient')],axis=1)
    pivot_F.reset_index(inplace=True)
    pivot_F = pivot_F.rename(columns={'index': column_name })
    return pivot_F
    
answer1 =  grouped_data('financial_class')
answer2 =  grouped_data('weekday')
print(answer1 , answer2)              #checking


  financial_class  waiting_ber_munets  number_of_patient
0       CORPORATE                46.0               6915
1             HMO                46.0               3738
2       INSURANCE                44.0               9931
3        MEDICARE                58.0                293
4         PRIVATE                40.0               9121      weekday  waiting_ber_munets  number_of_patient
0     Friday                42.0               4923
1     Monday                49.0               6982
2   Saturday                43.0               3010
3     Sunday                33.0               2549
4   Thursday                42.0               2673
5    Tuesday                42.0               5690
6  Wednesday                47.0               4171


In [32]:
fig = px.pie(answer1, values='number_of_patient', 
             names= 'financial_class' , hole=0.6 , width=600,height=600 , 
             template=template_style ,
             hover_data=['waiting_ber_munets'], 
             labels={'waiting_ber_munets':'the waiting time per m'})
fig.update_traces(textposition='outside', textinfo='percent+label')
fig.show()
# py.offline.plot(fig, filename='{}.html'.format(r"C:\" File Path "\\"),
#                 auto_open=False, image_width=1600,
#                         image_height=900)                     #import the plot to html file

## So the Q was: Dose the patient type affect the waiting time?

#### We can see that : 
#### - The 'INSURANCE' type of patient is (33.1%) of our patients and they are (9931 patients) and they waiting (44 min ) on average. 
#### - The 'PRIVATE' type of patient is (30.4%) of our patients and they are (9121 patients) and they waiting (40 min ) on average. 
#### - The 'CORPORATE' type of patient is (23.1%) of our patients and they are (6915 patients) and they wait (46 min ) on average. 
#### - The 'HMO' type of patient is (12.5%) of our patients and they are (3738 patients) and they waiting (46 min ) on average. 
#### - The 'MEDICARE' type of patient is (1%) of our patients and they are (293 patients) and they waiting (58 min ) on average. 

#### SO No the patient type does not give that big effect on the waiting time, BUT if you 'MEDICARE' type of patient your chance to wait long is higher.

_________________________________________

## Is there a specific type of patient waiting a long time?

#### Yes there is a specific type of patient waiting a long time, and that type is MEDICARE and they wait (58 min) on average.

___________________________________________________________________

## Are we too busy?
#### I'll do daily and hourly visualization to answer the Q, and build two heatmaps one for 'wait time' and the second for 'the number of patients, so lets make a function to make it easy

In [52]:
def grouped_data2(column_name):
    if column_name == 'completion_time':                                            # if column_name == 'completion_time' do the following
        answer = pp.pivot_table(df, index='hours',                                  # make a pivot_table and aggfunc 'count'
                                columns=['weekday'], aggfunc= 'count')
        answer.drop(answer.iloc[: , 7:], axis=1, inplace=True)                      # Drop the columns [: , 7:]
    elif column_name == 'waiting_ber_munets':                                       # if column_name == 'waiting_ber_munets' do the following
        answer = pp.pivot_table(df, index='hours',                                  # make a pivot_table and aggfunc 'mean'
                                columns=['weekday'] , aggfunc= 'mean').round(1)  
    else:
        return(column_name + " is not in the columns")                              # if the giving value is else print this statment
    
    answer = answer[column_name]
    answer = answer.fillna(0)
    answer = answer[['Sunday', 'Monday', 'Tuesday',                                 # sorting
                  'Wednesday', 'Thursday', 
                  'Friday', 'Saturday']] 
    return answer
answer3 =  grouped_data2('completion_time')
answer4 =  grouped_data2('waiting_ber_munets')
print(answer3,answer4)      #checking

weekday  Sunday  Monday  Tuesday  Wednesday  Thursday  Friday  Saturday
hours                                                                  
7           1.0     0.0      4.0        5.0       1.0     0.0       1.0
8         269.0   768.0    639.0      554.0     295.0   546.0     418.0
9         398.0   930.0    735.0      709.0     357.0   607.0     561.0
10        421.0   775.0    571.0      641.0     241.0   511.0     520.0
11        390.0   644.0    546.0      499.0     254.0   537.0     436.0
12        184.0   270.0    236.0      235.0     108.0   225.0     188.0
13         55.0   101.0     76.0       66.0      30.0    68.0      30.0
14        113.0   825.0    687.0      384.0     286.0   596.0     139.0
15         97.0   491.0    392.0      203.0     181.0   381.0     102.0
16         58.0   349.0    257.0      135.0     155.0   245.0      50.0
17         14.0    55.0     55.0       12.0      21.0    47.0      15.0
18        179.0   730.0    531.0      293.0     286.0   423.0   

C:\Users\abdh-\AppData\Local\Temp\ipykernel_25092\364743300.py:7: FutureWarning:

pivot_table dropped a column because it failed to aggregate. This behavior is deprecated and will raise in a future version of pandas. Select only the columns that can be aggregated.



In [53]:
fig3 = px.imshow(answer3,
                labels=dict(x="weekday", y="hours", color="number of paitant") , 
                aspect="auto", color_continuous_scale='tempo',
                template = template_style, title="Daily/hourly visualization",
                text_auto=True, width=700, height=700)
fig3.update_xaxes(side="top")
fig3.show()

#### I assumed that the highest number of patients = the longest wait time.
#### I noticed that there are times in the day when the number of patients is low, and these times are: (7 hour) and (13 hour) and (17 hour) and (21, and after)

In [54]:
fig4 = px.imshow(answer4,
                labels=dict(x="weekday", y="hours",
                            color="the waiting time per min") , 
                            aspect="auto", color_continuous_scale='tempo',
                            template = template_style,
                            text_auto=True, width=700, height=700)
fig4.update_xaxes(side="top")
fig4.show()

### Yes, we are too busy in the morning period and at 13 hour
#### In the first heatMap I thoughts it will be less wait time for patients at 7, 13, 17, 21, 22, and 23, But here I found that the rushed time for waiting started at 7 on(Tuesday, Wednesday, Thursday, Saturday) and reach the peak at 9, and decrease after, and start increasing at 13 and decreased after, and increased at 17 Until 20 after that it's decreasing, that means we are short in staff, but im curious about the 7 hour, why the wait time is hghi?

In [36]:
def morining(colmun_n):
    ddf = df[['entry_time' , 'post_consultation_time' , 'completion_time' , 'waiting_ber_munets']].sort_values(by= colmun_n ).head(10)
    return ddf

In [37]:
morining('entry_time').head(2)

entry_time post_consultation_time completion_time  \
570   0 days 07:53:25        0 days 08:29:46 0 days 08:49:16   
22590 0 days 07:55:06        0 days 08:22:58 0 days 08:31:59   

       waiting_ber_munets  
570                  56.0  
22590                37.0

#### It seems that the patients enter after 7:50 so maybe the clinic(OUTPATIENT) Starts at 8:00 

In [38]:
morining('post_consultation_time').head(2)

entry_time post_consultation_time completion_time  \
26482 0 days 07:58:59        0 days 08:07:41 0 days 08:07:57   
26483 0 days 08:02:59        0 days 08:08:34 0 days 08:11:32   

       waiting_ber_munets  
26482                 9.0  
26483                 9.0

#### it seems that im right the(OUTPATIENT) started at 8:07

### I'll build another chart to make sure that we are having a staff issue and nothing else

In [39]:
answer5 = grouped_data('hours')
link_size = [3,6]
fig = px.bar(x=answer5['hours'],
             y=answer5['waiting_ber_munets'],
             template= template_style,
             text_auto='.2s',
             labels={'x':'the hour',
                     'y':'the waiting time per m'}
             ).add_traces(
      px.line(answer5, x=answer5['hours'], text='number_of_patient',
             y=answer5['number_of_patient'],markers=True).update_traces(yaxis="y2",
             showlegend=True, line=dict(color = 'red', width=link_size[1]), name="number_of_patient").data)        
fig.update_layout(yaxis2={"side":"right", "overlaying":"y"})
fig.show()


#### Now we can see that (13 and 17) have an average wait time even though we have a lower amount of patients, so Yes it's a staffing issue

## So how can we fix this issue?
#### increasing our staffing at these hours (8 and 9,13 and 14,18) hours 

## What type of staff we need or where do we need them?

### Let's break this down, we have 4 types of timing in the data,
#### entry_time = entered the (OUTPATIENT)
#### post_consultation_time = when the doctor tells the patients to enter the clinic room
#### completion_time = when the patients exit the clinic room or the building
#### waiting_time = the entire wait time spend in the hospital
__________
we can extract more info from these like.
#### consultation_period = before entering to the doctor
#### process_period = talking to the doctor
#### consultation_perc = the % of time spend in consultation_period
#### process_perc = is the rest of the % of the consultation_perc time

In [40]:
df['consultation_period'] = df.post_consultation_time - df.entry_time
df['consultation_period'] = df.consultation_period.dt.seconds / 60          
df['consultation_period'] = df['consultation_period'].round(2)
# consultation_period = the amount of time after entr the (OUTPATIENT) and before the doctor see the patient

df['process_period'] = df.completion_time - df.post_consultation_time
df['process_period'] = df.process_period.dt.seconds / 60                     
df['process_period'] = df['process_period'].round(2)
# process_period = the amount of time the patients talking to the doctort + exite the (OUTPATIENT)

df['consultation_perc'] = df.consultation_period / df.waiting_ber_munets
df['consultation_perc'] = df['consultation_perc'].round(2)
# consultation_perc = the % of time spend in consultation_period

df['process_perc'] = 1 - df.consultation_perc
# process_perc = is the rest of the % of the consultation_perc time
df.head(1)

date medication_revenue lab_cost consultation_revenue doctor_type  \
0 2019-11-04            1183.22       10                20.17      ANCHOR   

  financial_class patient_type      entry_time post_consultation_time  \
0             HMO   OUTPATIENT 0 days 08:35:45        0 days 09:17:54   

  completion_time patient_id    waiting_time  waiting_ber_munets weekday  \
0 0 days 09:29:46     C10001 0 days 00:54:01                54.0  Monday   

   hours  consultation_period  process_period  consultation_perc  process_perc  
0      8                42.15           11.87               0.78          0.22

# How much the patients wait before the doctor can see them?

In [41]:
df.consultation_period.mean()

38.914371958130545

### The patients wait before the doctor can see them 38min in average

In [42]:
answer6 = df[['consultation_perc', 'process_perc']].mean()
periods = df[[ 'consultation_period', 'process_period']].mean().round(0)
print(answer6)

consultation_perc    0.882414
process_perc         0.117586
dtype: float64


In [43]:
fig = px.pie(answer6 , names= answer6.index, 
             template=template_style, 
             values= answer6,
             hover_name= periods.index,
             color=periods,
             labels={'color':'the waiting time per m '
                     })
fig.update_traces(textposition='outside',
                  textinfo='percent+label')
fig.show()

#### soo 88% of our patient time is gone for consultation_period whats means our patients spend 88% (average 39 min)of their time waiting for the doctor and spend just 11% talking to the doctor.

### as we can see the time to wait to see the doctor is limited by the number of doctors.

### The next Q was, What days of week are affected?

In [44]:
# Create Chart for the Daily
fig2 = px.bar(answer2,
             x='weekday',
             y='number_of_patient',
             color='waiting_ber_munets',
             labels={'waiting_ber_munets':'the waiting time per m'} ,
             color_continuous_scale=['green','yellow','red'],
             template = template_style,
             title = '<b>Daily visualization</b>')
# Display Plot
fig2.show()

##  To ansowe the Q, the affected day is Monday and Wednesday.
#### 'Monday' has a lot of patients(6982) and waiting time(49 min) (correlated)
#### 'Wednesday' has the average number of patients(4171) but high waiting time(47 min ) (not correlated)

# Summary: 
### There may be a possibility to add more medical staff during these times ((8 and 9,13 and 14,18) of the days), and focus on Monday and Wednesday.

# Actions:
### Determine if it makes financial sense to have an additional medical staffing at these times in summary section .